src/analytics/analytics-step-tracker.ts

```diff
@@ -0,0 +1,450 @@
+import { LineItemMap } from '../cart';
+import { Checkout, CheckoutService } from '../checkout';
+import { ShopperCurrency, StoreProfile } from '../config';
+import { Coupon } from '../coupon';
+import { Order } from '../order';
+import { ShippingOption } from '../shipping';
+
+import { AnalyticsTracker } from './analytics-tracker-window';
+import StepTracker from './step-tracker';
+
+export interface StepTrackerConfig {
+    checkoutSteps?: AnalyticStepType[];
+}
+
+export type AnalyticStepType = 'customer' | 'shipping' | 'billing' | 'payment';
+
+const ORDER_ITEMS_STORAGE_KEY = 'ORDER_ITEMS';
+
+export enum AnalyticStepId {
+    CUSTOMER = 1,
+    SHIPPING,
+    BILLING,
+    PAYMENT,
+}
+
+const ANALYTIC_STEPS: { [key: string]: AnalyticStepId } = {
+    customer: AnalyticStepId.CUSTOMER,
+    shipping: AnalyticStepId.SHIPPING,
+    billing: AnalyticStepId.BILLING,
+    payment: AnalyticStepId.PAYMENT,
+};
+
+export default class AnalyticsStepTracker implements StepTracker {
+    private checkoutStarted: boolean = false;
+    private completedSteps: { [key: string]: boolean } = {};
+    private viewedSteps: { [key in AnalyticStepId]?: boolean; } = {};
+    private analyticStepOrder: AnalyticStepType[] = [
+        'customer',
+        'shipping',
+        'billing',
+        'payment',
+    ];
+
+    constructor(
+        private checkoutService: CheckoutService,
+        private storage: StorageFallback,
+        private analytics: AnalyticsTracker,
+        { checkoutSteps }: StepTrackerConfig = {}
+    ) {
+        if (checkoutSteps !== undefined) {
+            this.analyticStepOrder = checkoutSteps;
+        }
+    }
+
+    trackCheckoutStarted(): void {
+        if (this.checkoutStarted) {
+            return;
+        }
+
+        const checkout = this.getCheckout();
+
+        if (!checkout) {
+            return;
+        }
+
+        const {
+            coupons,
+            grandTotal,
+            shippingCostTotal,
+            taxTotal,
+            cart: {
+                lineItems,
+                discountAmount,
+                id,
+            },
+        } = checkout;
+
+        const extraItemsData = this.saveExtraItemsData(id, lineItems);
+
+        this.analytics.track('Checkout Started', this.getTrackingPayload({
```

Comment:
Is it fine to send "Checkout Started" event again after a page refresh?

Reply:
that's current behaviour and I don't think merchants are complaining. I'm concerned that if I make a change here it could break the existing implementation. For example, if checkout is reloaded (after modifying the cart) the saved information could go out of sync.

Reply:
I see, that's a good point. I agree, it's better to keep it this way.